File: ios/Runner/PitchEngineChannel.swift
import Flutter

/// Bridges Dart's PitchEngine interface (lib/pitch_engine.dart) to one or more native
/// [PitchEngine] instances, keyed by an id Dart hands back on every call after `create` - mirrors
/// android/.../PitchEngineChannel.kt exactly, down to the wire format for `setInstrument`'s zone
/// data (SF2 parsing happens in Dart on both platforms).
///
/// This is an application-level channel (see AppDelegate.swift), not a registered FlutterPlugin -
/// there's no plugin package here, just a single app-owned native feature, so a plain
/// FlutterMethodChannel with this as its call handler is all that's needed.
final class PitchEngineChannel: NSObject {
    static let channelName = "com.arzaatri.dart_pitch_trainer/pitch_engine"

    private var engines: [Int: PitchEngine] = [:]
    private var nextId = 0

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "create" {
            let id = nextId
            nextId += 1
            engines[id] = PitchEngine()
            result(id)
            return
        }

        guard let args = call.arguments as? [String: Any], let id = args["id"] as? Int, let engine = engines[id] else {
            result(FlutterError(code: "UNKNOWN_ENGINE", message: "No PitchEngine for the given id", details: nil))
            return
        }

        switch call.method {
        case "start":
            engine.start()
            result(nil)
        case "stop":
            engine.stop()
            result(nil)
        case "release":
            engine.release()
            engines.removeValue(forKey: id)
            result(nil)
        case "updateFrequency":
            guard let frequency = args["frequency"] as? Double else {
                result(FlutterError(code: "BAD_ARGS", message: "Missing frequency", details: nil))
                return
            }
            engine.updateFrequency(frequency)
            result(nil)
        case "setVibratoEnabled":
            guard let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "BAD_ARGS", message: "Missing enabled", details: nil))
                return
            }
            engine.setVibratoEnabled(enabled)
            result(nil)
        case "setInstrument":
            guard let rawZones = args["zones"] as? [[String: Any]] else {
                result(FlutterError(code: "BAD_ARGS", message: "Missing zones", details: nil))
                return
            }
            engine.setInstrument(rawZones.map(Self.decodeZone))
            result(nil)
        case "isAudible":
            result(engine.isAudible())
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private static func decodeZone(_ map: [String: Any]) -> Sf2Zone {
        let pcmData = (map["pcm"] as? FlutterStandardTypedData)?.data ?? Data()
        // Composed byte-by-byte rather than bound directly to Int16 - Data's raw bytes aren't
        // guaranteed 2-byte aligned, so a direct bindMemory(to: Int16.self) would be undefined
        // behavior. Bytes arrive little-endian from Dart's Int16List.buffer.asUint8List().
        var pcm = [Int16](repeating: 0, count: pcmData.count / 2)
        pcmData.withUnsafeBytes { (rawBuffer: UnsafeRawBufferPointer) in
            for i in 0..<pcm.count {
                let lo = UInt16(rawBuffer[i * 2])
                let hi = UInt16(rawBuffer[i * 2 + 1])
                pcm[i] = Int16(bitPattern: lo | (hi << 8))
            }
        }
        return Sf2Zone(
            keyLo: map["keyLo"] as? Int ?? 0,
            keyHi: map["keyHi"] as? Int ?? 0,
            rootFrequency: map["rootFrequency"] as? Double ?? 440.0,
            gain: map["gain"] as? Double ?? 1.0,
            sampleRate: map["sampleRate"] as? Int ?? 44100,
            loop: map["loop"] as? Bool ?? false,
            loopStart: map["loopStart"] as? Int ?? 0,
            loopEnd: map["loopEnd"] as? Int ?? pcm.count,
            pcm: pcm,
            retriggerSeconds: map["retriggerSeconds"] as? Double ?? 0.0,
            vibratoCapable: map["vibratoCapable"] as? Bool ?? false
        )
    }
}

File: android/app/src/main/kotlin/com/arzaatri/dart_pitch_trainer/PitchEngineChannel.kt
package com.arzaatri.dart_pitch_trainer

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.atomic.AtomicInteger

/**
 * Bridges Dart's PitchEngine interface (lib/pitch_engine.dart) to one or more native
 * [PitchEngine] instances, keyed by an id Dart hands back on every call after `create`. Two
 * instances exist at once in practice - one each for the Tune and Guess panels - so they can
 * crossfade independently exactly like the original Android app's two AudioTracks.
 *
 * SF2 zones are parsed in Dart (lib/sound_font.dart) rather than natively, so `setInstrument`
 * receives already-resolved zone data (including precomputed rootFrequency/gain) instead of a
 * raw asset to parse.
 */
class PitchEngineChannel : MethodChannel.MethodCallHandler {
    private val engines = mutableMapOf<Int, PitchEngine>()
    private val nextId = AtomicInteger(0)

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method == "create") {
            val id = nextId.getAndIncrement()
            engines[id] = PitchEngine()
            result.success(id)
            return
        }

        val id = call.argument<Int>("id")
        val engine = id?.let { engines[it] }
        if (engine == null) {
            result.error("UNKNOWN_ENGINE", "No PitchEngine for id $id", null)
            return
        }

        when (call.method) {
            "start" -> {
                engine.start()
                result.success(null)
            }
            "stop" -> {
                engine.stop()
                result.success(null)
            }
            "release" -> {
                engine.release()
                engines.remove(id)
                result.success(null)
            }
            "updateFrequency" -> {
                val frequency = call.argument<Double>("frequency")!!
                engine.updateFrequency(frequency)
                result.success(null)
            }
            "setVibratoEnabled" -> {
                val enabled = call.argument<Boolean>("enabled")!!
                engine.setVibratoEnabled(enabled)
                result.success(null)
            }
            "setInstrument" -> {
                @Suppress("UNCHECKED_CAST")
                val rawZones = call.argument<List<Map<String, Any>>>("zones")!!
                engine.setInstrument(rawZones.map { decodeZone(it) })
                result.success(null)
            }
            "isAudible" -> result.success(engine.isAudible())
            else -> result.notImplemented()
        }
    }

    private fun decodeZone(map: Map<String, Any>): Sf2Zone {
        val pcmBytes = map["pcm"] as ByteArray
        val pcm = ShortArray(pcmBytes.size / 2)
        ByteBuffer.wrap(pcmBytes).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(pcm)
        return Sf2Zone(
            keyLo = (map["keyLo"] as Number).toInt(),
            keyHi = (map["keyHi"] as Number).toInt(),
            rootFrequency = (map["rootFrequency"] as Number).toDouble(),
            gain = (map["gain"] as Number).toDouble(),
            sampleRate = (map["sampleRate"] as Number).toInt(),
            loop = map["loop"] as Boolean,
            loopStart = (map["loopStart"] as Number).toInt(),
            loopEnd = (map["loopEnd"] as Number).toInt(),
            pcm = pcm,
            retriggerSeconds = (map["retriggerSeconds"] as Number).toDouble(),
            vibratoCapable = map["vibratoCapable"] as Boolean,
        )
    }

    /** Releases every still-open engine - called when the Flutter engine is torn down so no
     * AudioTrack playback threads are leaked past the activity's lifetime. */
    fun releaseAll() {
        engines.values.forEach { it.release() }
        engines.clear()
    }
}
